### PR TITLE
Add Data Channel receive from any source

### PR DESCRIPTION
**Pull request by VirrageS**

### CODE REVIEW

File: torch/csrc/distributed/Module.cpp [hunk 1]
```diff
@@ -177,7 +177,7 @@ PyObject* THDPModule_send(PyObject *_unused, PyObject *args)
   HANDLE_TH_ERRORS
   if (PyTuple_GET_SIZE(args) != 2 || !THPModule_isTensor(PyTuple_GET_ITEM(args, 0)) ||
         !THPUtils_checkLong(PyTuple_GET_ITEM(args, 1))) {
-    THPUtils_invalidArguments(args, "send", 1, "(tensor input, int dst_rank)");
+    THPUtils_invalidArguments(args, "isend", 1, "(tensor input, int dst_rank)");
```
Comment (apaszke) [inline torch/csrc/distributed/Module.cpp]: isn't this in `send`?

File: torch/distributed/collectives.py [hunk 2]
```diff
@@ -43,6 +43,10 @@ def send(tensor, dst_rank):
     return torch._C._dist_send(tensor, dst_rank)
 
 
+def recv(tensor):
```
Comment (apaszke) [inline torch/distributed/collectives.py]: I don't think you can do this, it's not C++ :wink:
```python
def recv(tensor, src_rank=None):
    if src_rank is None:
        return torch._C._dist_recv_any_source(tensor)
    return torch._C._dist_recv(tensor, src_rank)
```

File: torch/lib/THD/base/channels/DataChannelTCP.cpp [hunk 3]
```diff
@@ -673,6 +673,33 @@ void DataChannelTCP::receive(Scalar& data, int src_rank) {
 }
 
 
+void DataChannelTCP::receive(Tensor& data) {
+  auto request = _receive_worker.push([this, &data]{
+    std::unique_ptr<struct pollfd[]> events(new struct pollfd[this->_processes.size()]);
```
Comment (apaszke) [inline torch/lib/THD/base/channels/DataChannelTCP.cpp]: It would be good to cache this somewhere. It's small and constant, so it can be reused.

**Review by apaszke — CHANGES_REQUESTED** [review]
Only some minor things, otherwise LGTM

**Review by 0mp — APPROVED** [review]

File: torch/csrc/distributed/Module.cpp (hunk 1 above)
Comment (VirrageS) [reply torch/csrc/distributed/Module.cpp]: I have changed to `isend`, strange that git/github haven't noticed that.

**Review by apaszke — APPROVED** [review]

**Pull request merged**